### PR TITLE
Added nagios and jmx

### PR DESCRIPTION
**Pull request by kngu9**

### CODE REVIEW

File: charm/kafka/reactive/nagios.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        'object_name': 'kafka.controller:type=KafkaController,\
+name=ActiveControllerCount',
+        'description': 'Number of active controllers in cluster',
+        'crit': '!= 1'
+    }, {
+        'name': 'offline_partitions_count',
+        'object_name': 'kafka.controller:type=KafkaController,\
+name=OfflinePartitionsCount',
+        'description': 'Number of offline partitions in cluster',
+        'crit': '!= 1'
```
Comment (cmars) [inline charm/kafka/reactive/nagios.py]: Would this be crit if > 0? != 0 is fine as well, can't imagine negative offline partitions would be something you'd want to see either...

File: charm/kafka/reactive/nagios.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    }, {
+        'name': 'leader_election_rate',
+        'object_name': 'kafka.controller:type=ControllerStats,\
+name=LeaderElectionRateAndTimeMs',
+        'attribute': 'OneMinuteRate',
+        'description': 'Leader election rate and latency in milliseconds',
+        'warn': '>= 100',
```
Comment (cmars) [inline charm/kafka/reactive/nagios.py]: Might want to make these performance-related monitors configurable with defaults (this one and the ones below seem to be performance-related) rather than hard coded. As we operate kafka, we'll probably form opinions on where these should be set over time, and I could see different thresholds being used for different applications depending on their particular guarantees.

The checks above seem like basic cluster integrity checks, so it's fine to hard code those.

File: charm/kafka/reactive/autostart.py [hunk 3]
```diff
@@ -27,4 +27,4 @@ def autostart_service():
             hookenv.status_set('active', 'ready')
             return
 
-    hookenv.status_set('blocked', 'failed to start kafka; check syslog')
```
Comment (cmars) [inline charm/kafka/reactive/autostart.py]: I don't see where `status` is declared. Does this need to be reverted?

File: charm/kafka/reactive/kafka.py [hunk 4]
```diff
@@ -34,10 +34,22 @@ def disable_bootstrap_zk():
...
+@when('kafka.available', 'zookeeper.joined')
+@when_not('kafka.started', 'zookeeper.ready')
+def waiting_for_zookeeper_ready(zk):
+    hookenv.status_set('waiting', 'waiting for zookeeper to become ready')
+
+
 @hook('upgrade-charm')
 def upgrade_charm():
     remove_state('kafka.started')
-    remove_state('kafka.zk.disabled')
```
Comment (cmars) [inline charm/kafka/reactive/kafka.py:40]: I think we want to do both: remove `kafka.zk.disabled` as well as the nrpe_helper.installed. Also need to fix that flag name, it's not a zookeeper nrpe_helper flag...

**Review by cmars — COMMENTED** [review]
This is a great start!

File: charm/kafka/reactive/autostart.py (hunk 3 above)
Comment (kngu9) [reply charm/kafka/reactive/autostart.py]: Yep!

File: charm/kafka/reactive/tls.py [hunk 5]
```diff
@@ -119,7 +118,7 @@ def import_srv_crt_to_keystore():
                 log('importing pkcs12')
                 # import the pkcs12 into the keystore
                 check_call([
-                    KAFKA_KEYTOOL_PATH,
```
Comment (cmars) [inline charm/kafka/reactive/tls.py:122]: Why is this changed?

**Review by cmars — COMMENTED** [review]
Almost LGTM. I'm not entirely sure keytool is in the $PATH of the charm hook, see question below.

File: charm/kafka/reactive/kafka.py [hunk 6]
```diff
@@ -57,6 +46,17 @@ def waiting_for_zookeeper_ready(zk):
...
 
+@hook('upgrade-charm')
+def upgrade_charm():
+    remove_state('kafka.nrpe_helper.installed')
+    remove_state('kafka.started')
+
+
+@hook('config-changed')
+def config_changed():
+    remove_state('kafka.zk.disabled')
```
Comment (cmars) [inline charm/kafka/reactive/kafka.py]: We'll want to re-disable the embedded zookeeper on upgrade-charm, because the snap may have been reinstalled and that might re-enable the service. However, I don't think we need to do this on config-changed.

**Review by cmars — CHANGES_REQUESTED** [review]
See comments below on when the zookeeper service might need to be (re-)disabled. Conversely, we can remove `daemon: simple` from the zookeeper command in snapcraft.yaml, and stop installing it as a service altogether. I think I prefer that actually.

**Review by cmars — APPROVED** [review]
LGTM. I think layer:apt can be removed though.
